### PR TITLE
Add option to filter out tested CSRs using a REGEX on CSR name

### PR DESCRIPTION
**Pull request by PeterRugg**
This required refactoring of how we pass arguments into templates in general.

### CODE REVIEW

File: src/QuickCheckVEngine/Templates/Utils/FP.hs [hunk 1]
```diff
@@ -34,57 +34,19 @@
...
-
-fp_prologue_length :: ArchDesc -> Int
-fp_prologue_length = length . prologue_list
+fp_prologue = mconcat [ inst $ lui 1 2
+                      , csrs (unsafe_csrs_indexFromName "mstatus") 1
+                      , csrs (unsafe_csrs_indexFromName "fcsr") 0
+                      , readParams $ \p -> if has_f (archDesc p) || has_d (archDesc p)
+                                           then mconcat $ [inst $ fmv_w_x i 0 | i <- [0..4]]
+                                                       ++ [inst $ (if has_d (archDesc p) then fmv_d_x else fmv_w_x) i 0 | i <- [16..20]]
+                                           else mempty
```
Comment (arichardson) [inline src/QuickCheckVEngine/Templates/Utils/FP.hs:51]: Unrelated to this change, but I guess we should also skip writing to fcsr/mstatus when f&d are missing?

**Comment by arichardson** [comment]
It looks like this should work, but I wonder if it would be possible to add a builtin filter to remove *hpm* CSRs when ihpm is false?

File: src/QuickCheckVEngine/Templates/Utils/FP.hs (hunk 1 above)
Comment (PeterRugg) [reply src/QuickCheckVEngine/Templates/Utils/FP.hs:51]: It will! All the `csrs`-type helper templates now check the CSR isn't filtered out

Comment (PeterRugg) [reply src/QuickCheckVEngine/Templates/Utils/FP.hs:51]: Oh, hmm, sorry, not quite: most of the places that use this do the check of has_f || has_d as well though, so maybe they should get factored in.

**Comment by PeterRugg** [comment]
Yeah, I don't think it's plausible to ever get them to agree between two implementations that don't share a base (e.g. if you're checking some microarch optimisation hasn't changed anything). I would expect people to end up using the exclude-regex quite heavily, but it's slightly better (?) than the state of the world we have now where the templates have been tailored to avoid probing any registers that Flute/Toooba/Sail happen to treat differently in implementation-defined ways. At least the user knows what they're excluding?

**Pull request merged**